src/color.rs

```diff
@@ -150,335 +97,606 @@ impl<'de> Deserialize<'de> for RGBA {
         D: Deserializer<'de>,
     {
         let (r, g, b, a) = Deserialize::deserialize(deserializer)?;
-        Ok(RGBA::new(r, g, b, a))
+        Ok(Self::new(r, g, b, a))
     }
 }
 
-impl ToCss for RGBA {
-    fn to_css<W>(&self, dest: &mut W) -> fmt::Result
-    where
-        W: fmt::Write,
-    {
-        let has_alpha = self.alpha != OPAQUE;
+// NOTE: `RGBA` should not implement `ToCss` because it is an internal primitive that does not directly represent a CSS value.
 
-        dest.write_str(if has_alpha { "rgba(" } else { "rgb(" })?;
-        self.red.to_css(dest)?;
-        dest.write_str(", ")?;
-        self.green.to_css(dest)?;
-        dest.write_str(", ")?;
-        self.blue.to_css(dest)?;
+#[inline]
+fn f32_to_f64(f: f32) -> f64 {
+    const PRECISION: f64 = 1e6;
+    (f64::from(f) * PRECISION).round() / PRECISION
+}
 
-        serialize_alpha(dest, self.alpha, true)?;
+/// <hue>
+/// https://w3c.github.io/csswg-drafts/css-color-4/#hue-syntax
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub struct Hue {
+    /// The value as a number of degrees.
+    pub degrees: f32,
+}
 
-        dest.write_char(')')
+impl Hue {
+    /// Construct a new Hue from a number of degrees.
+    pub fn new(degrees: f32) -> Self {
+        Self { degrees }
     }
 }
 
-// NOTE: LAB and OKLAB is not declared inside the [impl_lab_like] macro,
-// because it causes cbindgen to ignore them.
+/// <alpha-value>
+/// https://w3c.github.io/csswg-drafts/css-color-4/#alpha-syntax
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub struct AlphaValue {
+    /// The value as a number in the range of 0 to 1.
+    pub number: f32,
+}
+
+impl AlphaValue {
+    /// Construct a new AlphaValue from a number in the range of 0 to 1.
+    pub fn new(number: f32) -> Self {
+        Self { number }
+    }
+}
 
-/// Color specified by lightness, a- and b-axis components.
+// A set of coordinates in the Lab rectangular coordinate system.
 #[derive(Clone, Copy, PartialEq, Debug)]
-#[repr(C)]
-pub struct Lab {
-    /// The lightness component.
-    pub lightness: f32,
-    /// The a-axis component.
-    pub a: f32,
-    /// The b-axis component.
-    pub b: f32,
-    /// The alpha component.
-    pub alpha: f32,
+pub struct LabCoords {
+    pub lightness: Option<f32>,
+    pub a: Option<f32>,
+    pub b: Option<f32>,
+    pub alpha: Option<AlphaValue>,
+}
```

Comment:
For the record, I don't think using `Option<f32>` is going to work great for us in terms of representing `none` coordinates. First, it grows the color structs quite considerably.

Second, it's not possible to use it from C/C++, which is a problem for Firefox integration.

Can we do something like this instead?

```
#[repr(transparent)]
pub struct OptionalColorCoordinate(pub f32);

impl OptionalColorCoordinate {
    fn none() -> Self { Self(std::f32::NAN) }
    fn get() -> Option<f32> {
        if self.0.is_nan() { None } else { Some(self.0) }
    }
}
```

Or something along those lines? Same for `Option<AlphaValue>` etc.

Reply:
@emilio Can you go into more detail about the problems this design causes? I don't like the idea of abusing NaN to capture a concept independent of the float value.

@tiaanl Do you have any thoughts on this?

Reply:
I outlined them above:

 * Bigger structs are bad for CSS parsing performance (since they're returned by value quite a lot) and memory usage (if they're stored for long periods of time, which in the case of Firefox they are). Colors are rather common, so it'd be great not to regress parsing performance and memory usage any more than necessary. A struct of four `Option<f32>`s is twice as big as `[f32; 4]`, see [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=519921cb0464b5d02b90d97e0b319601), so if there's a not-too-complicated way to halve the size of a color, we should definitely do that.
 * Additionally, we'd love to be able to avoid converting between different color representations between C/C++ and Rust. `Option<>` is not ffi-safe, in general, so we would need to convert colors after parsing to another representation in Firefox if we used it. We _could_ do it if absolutely needed, but if we can avoid it (since it's just boilerplate and potentially duplicated code we'd have to maintain), it'd be great.

Using `NaN` was an idea to deal with both of thos. It allows to encode `none` without wasting extra bits, and it shouldn't make it particularly harder to use from Rust if we do something like what I suggested above, I think? But I'm open to other options.

In an ideal world, there'd be something like a `NotNaN<T: Float>` in the standard library that guarantees the right range so that `Option<NotNaN<f32>>` is guaranteed to be 32 bits, just like `Option<Box<T>>` is guaranteed to be the size of a pointer as long as `T` is `Sized`. That would address both and would be perhaps slightly more idiomatic, but given that doesn't exist, rolling our own as described above seems reasonable?

Another option would be to use something like fixed-point representation where 15 or 31 bits are used for the actual number, and one bit is used for `none`, but that's also a bit more code, and at that point we need to make rather arbitrary precision choices, which would be great to avoid if possible. It's one of the things we could consider doing (using f15 + a bit for `none`) if using f32 for all this becomes prohibitively expensive in terms of memory usage (since it could halve sizes compared to f32 representation). We use that kind of trick in some other places in Firefox like for font-style angles (see [this](https://searchfox.org/mozilla-central/rev/f24503d2f5d8b7e4341be8bc20d94a04a6b223d6/servo/components/style/values/computed/font.rs#845-854)), but again, I'd prefer to avoid it if possible.

Anyways, hopefully that clarifies my concerns with this approach?

Reply:
Ah, yes, thank you for that context. I think I understand the issue a bit better now.

I'll try my best to accommodate your request, but one of the things I intended to include in the background explanation that I have yet to write is that I'd like to consider optimizations to be out of scope as much as possible. These changes are unwieldy enough already, and I'd prefer to not add further delay while I fiddle with the idiosyncrasies of Rust; memory management is not my area of expertise. Do we have tests that would be tracking regressions of this nature? And if not, can we commit to adding them in the future?

I'd also like to make explicit one point that I think is already implicit in your comment: We want to be sure to differentiate between optimizations (and other things) that are broadly applicable and those that are Firefox-specific. The former are definitely appropriate for this repo, but the latter probably aren't.

Reply:
> Do we have tests that would be tracking regressions of this nature?

Sure, we do have CSS parsing benchmarks for Firefox like [this](https://searchfox.org/mozilla-central/source/layout/style/test/gtest/StyloParsingBench.cpp), plus other performance tests that are monitored regularly.

That said, those obviously benchmark more than color parsing, and need a Firefox checkout plus incorporating your changes into Firefox which I don't think is something reasonable to ask of you.

But it should be rather trivial to add a few benchmarks for color parsing in `src/tests.rs` like we have for `data:` uris etc?

> We want to be sure to differentiate between optimizations (and other things) that are broadly applicable and those that are Firefox-specific. The former are definitely appropriate for this repo, but the latter probably aren't.

Sure, I agree. Firefox has the extra requirements of dealing with C++ and thus needing ffi-safe types etc. I'd be ok merging a change that requires Firefox to support ffi on top of cssparser (even though I think it'd be unfortunate).

However general parsing performance should be applicable to all other rust crates that depend on `cssparser`. Firefox might have stricter performance requirements than some other programs, but I don't think I'd be fine merging a change with performance characteristics that don't work for Firefox.